File: sdk/python/src/cyt_indexer/paths.py
"""Path helpers and runtime config (defaults and overrides live in Rust PathConfig)."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from cyt_indexer._native import (
    collect_enums as _collect_enums,
)
from cyt_indexer._native import (
    configure_path_constants as _configure_path_constants_native,
)
from cyt_indexer._native import (
    get_root_tool_key as _get_root_tool_key,
)
from cyt_indexer._native import (
    path_builder_memory_only as _path_builder_memory_only,
)
from cyt_indexer._native import (
    path_catalog_prefix as _path_catalog_prefix,
)
from cyt_indexer._native import (
    path_decomposed_prefix as _path_decomposed_prefix,
)
from cyt_indexer._native import (
    path_decomposed_root as _path_decomposed_root,
)
from cyt_indexer._native import (
    path_default_catalog_dir as _path_default_catalog_dir,
)
from cyt_indexer._native import (
    path_json_ext as _path_json_ext,
)
from cyt_indexer._native import (
    path_md_ext as _path_md_ext,
)
from cyt_indexer._native import (
    path_write_catalog_prune as _path_write_catalog_prune,
)
from cyt_indexer._native import (
    to_decomposed_key as _to_decomposed_key,
)
from cyt_indexer._native import (
    tool_id_from_decomposed_rel as _tool_id_from_decomposed_rel,
)


def configure_path_constants(
    *,
    md_ext: str,
    json_ext: str,
    decomposed_prefix: str,
    decomposed_root: str | Path,
    catalog_prefix: str,
    builder_memory_only: bool,
    default_catalog_dir: str | Path,
    write_catalog_prune: bool,
) -> None:
    """Push host app overrides into native PathConfig (Rust defaults when not called)."""
    _configure_path_constants_native(
        md_ext,
        json_ext,
        decomposed_prefix,
        str(decomposed_root),
        catalog_prefix,
        str(default_catalog_dir),
        (builder_memory_only, write_catalog_prune),
    )


def md_ext() -> str:
    return _path_md_ext()


def json_ext() -> str:
    return _path_json_ext()


def decomposed_prefix() -> str:
    return _path_decomposed_prefix()


def decomposed_root() -> str:
    return _path_decomposed_root()


def catalog_prefix() -> str:
    return _path_catalog_prefix()


def default_catalog_dir() -> str:
    return _path_default_catalog_dir()


def builder_memory_only() -> bool:
    return _path_builder_memory_only()


def write_catalog_prune() -> bool:
    return _path_write_catalog_prune()


def to_decomposed_key(file_path: str) -> str | None:
    return _to_decomposed_key(file_path)


def tool_id_from_decomposed_rel(rel_path: str) -> str:
    return _tool_id_from_decomposed_rel(rel_path)


def get_root_tool_key(file_path: str) -> str | None:
    return _get_root_tool_key(file_path)


def collect_enums(schema: dict[str, Any]) -> list[dict[str, Any]]:
    return list(_collect_enums(schema))

File: sdk/go/cytindexer/paths.go
package cytindexer

import (
	"path/filepath"
	"strings"
	"sync"
)

// PathConfig holds SDK runtime defaults for paths and catalog I/O.
type PathConfig struct {
	JSONExt            string
	MDExt              string
	DecomposedPrefix   string
	DecomposedRoot     string
	CatalogPrefix      string
	BuilderMemoryOnly  bool
	DefaultCatalogDir  string
	WriteCatalogPrune  bool
}

// DefaultPathConfig returns the default path configuration.
func DefaultPathConfig() PathConfig {
	return PathConfig{
		JSONExt:           ".json",
		MDExt:             ".md",
		DecomposedPrefix:  "schemas/decomposed/",
		DecomposedRoot:    "schemas/decomposed",
		CatalogPrefix:     "catalog",
		BuilderMemoryOnly: false,
		DefaultCatalogDir: "catalog",
		WriteCatalogPrune: true,
	}
}

var (
	pathMu     sync.RWMutex
	pathConfig = DefaultPathConfig()
)

// ConfigurePaths overrides path configuration.
func ConfigurePaths(cfg PathConfig) {
	pathMu.Lock()
	pathConfig = cfg
	pathMu.Unlock()
}

// PathSnapshot returns a copy of the current path configuration.
func PathSnapshot() PathConfig {
	pathMu.RLock()
	defer pathMu.RUnlock()
	return pathConfig
}

func jsonExt() string     { return PathSnapshot().JSONExt }
func mdExt() string       { return PathSnapshot().MDExt }
func decomposedPrefix() string { return PathSnapshot().DecomposedPrefix }
func decomposedRoot() string   { return PathSnapshot().DecomposedRoot }

// JSONExt returns the configured JSON file extension.
func JSONExt() string { return jsonExt() }

// MDExt returns the configured markdown file extension.
func MDExt() string { return mdExt() }

// DecomposedPrefix returns the decomposed schema directory prefix.
func DecomposedPrefix() string { return decomposedPrefix() }

// DecomposedRoot returns the decomposed schema root path.
func DecomposedRoot() string { return decomposedRoot() }

// CatalogPrefix returns the catalog prefix string.
func CatalogPrefix() string { return PathSnapshot().CatalogPrefix }

// DefaultCatalogDir returns the default on-disk catalog directory.
func DefaultCatalogDir() string { return PathSnapshot().DefaultCatalogDir }

// WriteCatalogPrune returns whether stale catalog files should be pruned on write.
func WriteCatalogPrune() bool { return PathSnapshot().WriteCatalogPrune }

// ToDecomposedKey normalizes a file path to a decomposed catalog key.
func ToDecomposedKey(filePath string) (string, bool) {
	clean := filepath.ToSlash(filepath.Clean(filePath))
	parts := strings.Split(clean, "/")
	for i := 0; i < len(parts)-1; i++ {
		if parts[i] == "schemas" && parts[i+1] == "decomposed" {
			return strings.Join(parts[i:], "/"), true
		}
	}
	return "", false
}

// ToolIDFromDecomposedRel extracts the tool id from a decomposed relative path.
func ToolIDFromDecomposedRel(relPath string) string {
	cfg := PathSnapshot()
	rel := relPath
	if strings.HasPrefix(rel, cfg.DecomposedPrefix) {
		rel = strings.TrimPrefix(rel, cfg.DecomposedPrefix)
	}
	parts := strings.Split(filepath.ToSlash(rel), "/")
	if len(parts) == 0 {
		base := filepath.Base(rel)
		return strings.TrimSuffix(base, cfg.JSONExt)
	}
	first := parts[0]
	if strings.HasSuffix(first, cfg.JSONExt) {
		return strings.TrimSuffix(first, cfg.JSONExt)
	}
	return first
}

// GetRootToolKey returns the root decomposed key for a catalog file path.
func GetRootToolKey(filePath string) (string, bool) {
	cfg := PathSnapshot()
	key, ok := ToDecomposedKey(filePath)
	if !ok {
		return "", false
	}
	rel, err := filepath.Rel(cfg.DecomposedRoot, key)
	if err != nil || rel == "." || rel == "" {
		return "", false
	}
	parts := strings.Split(filepath.ToSlash(rel), "/")
	if len(parts) == 1 && strings.HasSuffix(parts[0], cfg.JSONExt) {
		return key, true
	}
	toolID := parts[0]
	return cfg.DecomposedPrefix + toolID + cfg.JSONExt, true
}

// CollectEnums recursively collects enum values from a JSON schema.
func CollectEnums(schema any) []any {
	var found []any
	collectEnumsInner(schema, &found)
	return found
}

func collectEnumsInner(node any, found *[]any) {
	switch n := node.(type) {
	case map[string]any:
		if items, ok := AsArray(n["enum"]); ok {
			*found = append(*found, items...)
		}
		for _, val := range n {
			if _, isObj := AsObject(val); isObj {
				collectEnumsInner(val, found)
			} else if _, isArr := AsArray(val); isArr {
				collectEnumsInner(val, found)
			}
		}
	case []any:
		for _, item := range n {
			if _, isObj := AsObject(item); isObj {
				collectEnumsInner(item, found)
			} else if _, isArr := AsArray(item); isArr {
				collectEnumsInner(item, found)
			}
		}
	}
}
